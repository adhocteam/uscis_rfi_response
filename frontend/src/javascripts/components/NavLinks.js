import React from "react";
import { Link } from "react-router-dom";

import history from "../services/history";
import { isLoggedIn } from "../helpers/session";

const allLinks = [
  {
    url: "/login",
    text: "Log In",
    show_when_logged_out: true,
  },
  {
    url: "/logout",
    text: "Log Out",
    show_when_logged_in: true,
  },
  {
    url: "/",
    text: "Upload Image",
    show_when_logged_in: true,
    show_when_logged_out: true,
  },
  {
    url: "/request",
    text: "Request an Upload",
    show_when_logged_in: true,
  },
  {
    url: "/review",
    text: "Review Submissions",
    show_when_logged_in: true,
  },
];

class NavLinks extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      loggedIn: isLoggedIn(),
    };
    // on route change, update state
    history.listen(() => {
      this.setState({ loggedIn: isLoggedIn() });
    });
  }

  render() {
    const { loggedIn } = this.state;
    return (
      <div className="ds-l-col--12 ds-l-md-col--8 navigation">
        <ul className="ds-u-margin--0 ds-u-margin-top--1 ds-u-margin-bottom--1 ds-u-md-margin-top--3 ds-u-md-margin-bottom--0 ds-u-padding--0 ds-u-md-text-align--right">
          {allLinks
            .filter(link => {
              return loggedIn
                ? link.show_when_logged_in
                : link.show_when_logged_out;
            })
            .map((link, i) => {
              return (
                <li
                  key={i}
                  className="ds-u-display--inline-block  ds-u-margin-right--2"
                >
                  <Link
                    to={link.url}
                    className="ds-u-color--white ds-u-font-weight--bold"
                  >
                    {link.text}
                  </Link>
                </li>
              );
            })}
        </ul>
      </div>
    );
  }
}

export default NavLinks;
