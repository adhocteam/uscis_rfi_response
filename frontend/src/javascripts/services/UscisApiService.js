import history from "./history";

const updateIfPresent = (headers, headerName, sessionName) => {
  const val = headers.get(headerName);
  if (val) {
    sessionStorage.setItem(sessionName, val);
  }
};

const updateTokens = headers => {
  updateIfPresent(headers, "access-token", "token");
  updateIfPresent(headers, "expiry", "expiry");
  updateIfPresent(headers, "uid", "uid");
  updateIfPresent(headers, "client", "client");
};

const authedRequest = (path, settings, error) =>
  fetch(`${BASE_URL}${path}`, {
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
      "access-token": sessionStorage.getItem("token"),
      client: sessionStorage.getItem("client"),
      uid: sessionStorage.getItem("uid"),
      expiry: sessionStorage.getItem("expiry"),
    },
    ...settings,
  }).then(resp => {
    if (resp.ok) {
      updateTokens(resp.headers);
      return resp.json();
    }
    if (resp.status === 401) {
      history.replace(`/login?next=${history.location.pathname}`);
    }
    throw new Error(error);
  });

const UscisApiService = {
  getAdmin: () => authedRequest("/admin", {}, "Failed to get admin."),

  // TODO: error handling
  getSignedUrl: (user_id, image_name, image_type) => {
    return fetch(`${BASE_URL}/submissions/presigned_url`, {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        submission_id: submission_id,
        image_name: image_name,
        image_type: image_type,
      }),
    }).then(resp => {
      if (resp.status === 200) {
        return resp.json();
      }
    });
  },

  filterSubmissions: status => {
    return authedRequest(
      `/submissions/filter?status=${status}`,
      {},
      "Failed to get submissions."
    );
  },

  getSubmission: id =>
    authedRequest(`/submissions/${id}`, {}, `Failed to get submission ${id}.`),

  getSubmissions: () =>
    authedRequest("/submissions", {}, "Failed to get submissions."),

  login: (email, password) => {
    return fetch(`${BASE_URL}/auth/sign_in`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Accept: "application/json",
      },
      body: JSON.stringify({
        email: email,
        password: password,
      }),
    }).then(resp => {
      if (resp.ok) {
        updateTokens(resp.headers);
        return resp.json();
      } else {
        throw new Error("Failed to log in.");
      }
    });
  },

  updateSubmission: ({ id, notes, status }) => {
    authedRequest(
      `/submissions/${id}`,
      {
        method: "PUT",
        body: JSON.stringify({ notes, status }),
      },
      `Failed to update submission ${id}.`
    );
  },

  createUser: (name, dob, email, street1, street2, city, state, zip) => {
    return authedRequest(`submissions/new_upload`, {
      method: "POST",
      body: JSON.stringify({
        name: name || "",
        dob: dob || "",
        email: email || "",
        street1: street1 || "",
        street2: street2 || "",
        city: city || "",
        state: state || "",
        zip: zip || "",
      }),
    });
  },
};

export default UscisApiService;
