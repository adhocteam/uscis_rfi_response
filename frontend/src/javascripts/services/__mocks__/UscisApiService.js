const UscisApiService = {
  login: (email, password) => {
    sessionStorage.setItem("token", "FAKE-ACCESS-TOKEN");
    sessionStorage.setItem("client", "FAKE-CLIENT-ID");
    sessionStorage.setItem("uid", email);
    sessionStorage.setItem("expiry", Date.now() + 1209600000); //two weeks
    return Promise.resolve();
  },
};

export default UscisApiService;
