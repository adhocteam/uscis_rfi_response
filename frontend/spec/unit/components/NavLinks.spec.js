jest.mock("services/UscisApiService");

import React from "react";
import { shallow } from "enzyme";
import NavLinks from "components/NavLinks.js";
import UscisApiService from "services/UscisApiService.js";

test("renders certain links when logged out", () => {
  const wrapper = shallow(<NavLinks />);
  expect(wrapper).toMatchSnapshot();
});

test("renders certain links when logged in", () => {
  UscisApiService.login("fake.admin@adhocteam.us", "thisisafakepassword");
  const wrapper = shallow(<NavLinks />);
  expect(wrapper).toMatchSnapshot();
});
