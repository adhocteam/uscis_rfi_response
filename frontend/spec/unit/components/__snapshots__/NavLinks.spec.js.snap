// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders certain links when logged in 1`] = `
<div
  className="ds-l-col--12 ds-l-md-col--8 navigation"
>
  <ul
    className="ds-u-margin--0 ds-u-margin-top--1 ds-u-margin-bottom--1 ds-u-md-margin-top--3 ds-u-md-margin-bottom--0 ds-u-padding--0 ds-u-md-text-align--right"
  >
    <li
      className="ds-u-display--inline-block  ds-u-margin-right--2"
      key="0"
    >
      <Link
        className="ds-u-color--white ds-u-font-weight--bold"
        replace={false}
        to="/logout"
      >
        Log Out
      </Link>
    </li>
    <li
      className="ds-u-display--inline-block  ds-u-margin-right--2"
      key="1"
    >
      <Link
        className="ds-u-color--white ds-u-font-weight--bold"
        replace={false}
        to="/"
      >
        Upload Image
      </Link>
    </li>
    <li
      className="ds-u-display--inline-block  ds-u-margin-right--2"
      key="2"
    >
      <Link
        className="ds-u-color--white ds-u-font-weight--bold"
        replace={false}
        to="/request"
      >
        Request an Upload
      </Link>
    </li>
    <li
      className="ds-u-display--inline-block  ds-u-margin-right--2"
      key="3"
    >
      <Link
        className="ds-u-color--white ds-u-font-weight--bold"
        replace={false}
        to="/review"
      >
        Review Submissions
      </Link>
    </li>
  </ul>
</div>
`;

exports[`renders certain links when logged out 1`] = `
<div
  className="ds-l-col--12 ds-l-md-col--8 navigation"
>
  <ul
    className="ds-u-margin--0 ds-u-margin-top--1 ds-u-margin-bottom--1 ds-u-md-margin-top--3 ds-u-md-margin-bottom--0 ds-u-padding--0 ds-u-md-text-align--right"
  >
    <li
      className="ds-u-display--inline-block  ds-u-margin-right--2"
      key="0"
    >
      <Link
        className="ds-u-color--white ds-u-font-weight--bold"
        replace={false}
        to="/login"
      >
        Log In
      </Link>
    </li>
    <li
      className="ds-u-display--inline-block  ds-u-margin-right--2"
      key="1"
    >
      <Link
        className="ds-u-color--white ds-u-font-weight--bold"
        replace={false}
        to="/"
      >
        Upload Image
      </Link>
    </li>
  </ul>
</div>
`;
